### ISSUE TITLE
✨ Compute "final decision" and "current assessment step" for Pool Candidates on backend

### ISSUE THREAD

**Issue by tristan-orourke** [GCTC-NTGC/gc-digital-talent#8966]
## ✨ Feature

While a Pool Candidate is in assessment, we can consider which steps have been completely assessed and which require further assessments. A step could also be unsucceessful.

It would be useful to compute this on the backend via database queries, since this would allow us to sort candidates by "Current step" and simplify some frontend components.

This will enable #8960.

## 🕵️ Details

Assessment step status is determined by:
```
Essential requirements = essential skills AND education requirements
Optional requirements = asset skills
Assessment Step Status =
  IF any essential requirements unsuccessful THEN "unsuccessful"
  IF any requirements unassessed THEN "to assess"
  IF essential requirements include at least one hold THEN "hold"
  IF all essential requirements successful THEN "success"

Also, on the step tracker, a step effectively only has the "to assess" status if all previous steps are "hold" or "success".
```

The "current step" is 
- null (not applicable) if all steps are successful or any steps are unsuccessful
- OR the number of the first "to assess"/"hold" step

### Sort order
The order of statuses should be as following:

Qualified
Qualified pending decision
To assess step 1
To assess step 2
....
Disqualified
Disqualified pending decision
Removed

## 🎨 Design File

Include a link to the design file (if it exists).

### 🧑‍🎨 Designer

@Jerryescandon 

## 📸 Screenshot

Add a screenshot of the new feature (if possible).

## 🙋‍♀️ Proposed Implementation

Look at how we implemented Skill counts for pool candidate search. We can do something like that, I think.

For sorting, either sort like the "skill count" column, or an additional generated column for sorting.

## 🌎 Localization

(optional) Provide any new copy along with translations available.

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the feature was properly implemented.

- [ ] When getting one or more pool candidates from the api, you can request a `currentStep: number` field
- [ ] New enum to represent: Final decision status 
- [ ] Way of sorting that includes both final decision and current step
- [ ] You can get the final decision status from the api

🛑 Blockers
```[tasklist]
### Blocked by
- [ ] #10110
```

**Comment by gobyrne** [comment]
@tristan-orourke need a new ticket to represent the refactoring

**Comment by vd1992** [comment]
Blocking on https://github.com/GCTC-NTGC/gc-digital-talent/issues/10110
As I think quality test data will be essential for this issue, both in completion and review. Seeding invalid states will be a thorn otherwise

**Comment by vd1992** [comment]
@petertgiles 
Per today's scrum, yes I do have questions about the issue from having looked at it before. 

### Performance

`finalDecision` and `currentStep` are likely to have a lot of overlap in their computation and involve querying multiple tables like 
`poolCandidate.assessmentResults` and `poolCandidate.pool.assessmentSteps.poolSkills` (need steps and attached skills)

Grouping them into one object

```
screeningAssessmentStatus => [
 finalDecision: FinalDecisionEnum,
 currentStep: Int | null
]
```

returned by a single accessor function could offer better performance. Though, you might not be able to or will struggle with sorting this? Or maybe computing finalDecision and currentStep separately doesn't make a performance difference when you load a table with 100 or more candidates?

### Implementation of sorting

**Comment by esizer** [comment]
I have been trying to mull this over to think about how we can do it and I was wondering if we could do it with event listeners that update either a single database column on `PoolCandidate` or a relationhip 🤔 

I have some ides laid out in some rough code for input on if I am headed down the correct path or not.

```php
<?php


class Migration
{
    public function up() {
        Schema::table('pool_candidate', function(Blueprint $table) {
            $table->jsonb('assessment_results')->nullable();
        });
    }
}

class AssessmentStepResult extends Model
{
    protected $dispatchesEvents = [
        'saved' => AssessmentResultSaved::class,
    ];
}

class AssessmentResultSaved
{
    use Dispatchable;

    public function __construct(public AssessmentStepResult $result){}
}

class ComputeDecisionAndStep
{

    public function handle(AssessmentResultSaved $event)
    {

        $poolCandidate = $event->pool_candidate;
        $steps = $poolCandidate->pool->assessmentSteps;

        $decisions = $this->decisionPerStep(
            $poolCandidate->results,
            $steps
        );

        $currentStep = $this->currentStep($decisions, $steps);

        $poolCandidate->assessment_results = [
            'decisions' => $decisions, // This could also be a related table if it makes things easier (pool_candidate_decisions)
            'current_step' => $currentStep
        ];

        $poolCandidate->save();
    }
}

// We could then get candidates by current step

$query->where('assessment_results->current_step', '>=', $assessmentStep->step_order);
```

**Comment by esizer** [comment]
> Like a "memo" or "calculated" suffix so it's clear that they shouldn't be updated directly.

Yeah, that is a good idea. I feel like without that we could easily make some mistakes :sweat_smile: 

> I'm not sure I like the extra model for this. I feel like this memoization logic should be an implementation detail and not part of the model set.

Yeah, I wasn't really happy with the extra model either so I only added it as an option :sweat_smile: Thanks for confirming my distaste for that approach :stuck_out_tongue:

**Comment by tristan-orourke** [comment]
These are good ideas here, but I also spent some time trying to write the SQL statement which would do this at the database level. Just to see if I could, and so we have something to compare other solutions to. It's not quite done yet, but almost!

```
SELECT pool_candidates.id,
  essential_results.assessment_step_id,
  essential_results.assessment_decisions,
  steps.needs_education,
  steps.essential_skill_count,
  (steps.needs_education + steps.essential_skill_count) as "total_criteria_count",
  array_length(essential_results.assessment_decisions, 1) as "assessed_criteria_count",
  array_length(essential_results.assessment_decisions, 1) >= (steps.needs_education + steps.essential_skill_count) as "all_criteria_assessed",
FROM pool_candidates
LEFT JOIN (
  SELECT assessment_results.pool_candidate_id,
    assessment_results.assessment_step_id, 
    array_agg(assessment_results.assessment_decision) as "assessment_decisions",
    array_agg(pool_skill.type) as skill_types,
    array_agg(skills.name->>'en') as skill_names
  FROM assessment_results
  LEFT JOIN pool_skill ON pool_skill.id = assessment_results.pool_skill_id
  LEFT JOIN skills ON skills.id = pool_skill.skill_id
  WHERE (
    assessment_result_type = 'EDUCATION'
    OR pool_skill.type = 'ESSENTIAL'
  )
  GROUP BY assessment_results.pool_candidate_id, assessment_results.assessment_step_id
) essential_results ON pool_candidates.id = essential_results.pool_candidate_id
JOIN (
  SELECT 
    assessment_steps.id as "step_id",
    CASE WHEN assessment_steps.type = 'APPLICATION_SCREENING' THEN 1 ELSE 0 END as needs_education,
    array_length(array_agg(pool_skill.id), 1) as essential_skill_count,
    array_agg(pool_skill.type) as skill_types,
    array_agg(skills.name->>'en') as skill_names
  FROM assessment_steps
  LEFT JOIN assessment_step_pool_skill ON assessment_step_pool_skill.assessment_step_id = assessment_steps.id
  LEFT JOIN pool_skill ON pool_skill.id = assessment_step_pool_skill.pool_skill_id
  LEFT JOIN skills ON skills.id = pool_skill.skill_id
  WHERE pool_skill.type = 'ESSENTIAL'
  GROUP BY assessment_steps.id
) steps ON step_id = essential_results.assessment_step_id;
```
